File: src/services/egress-monitor.ts
/**
 * 出口 IP 监控服务
 */

import { invoke } from '@tauri-apps/api/core'

import type {
  EgressMonitorConfig,
  EgressMonitorStats,
  EgressIpProbeResult,
} from '@/services/coordinator'

export type {
  EgressMonitorConfig,
  EgressMonitorStats,
  EgressIpProbeResult,
  EgressIpChangeEvent,
  RebindStrategyType,
} from '@/services/coordinator'

/**
 * 获取出口监控配置
 */
export async function egressMonitorGetConfig(): Promise<EgressMonitorConfig> {
  return await invoke<EgressMonitorConfig>('egress_monitor_get_config')
}

/**
 * 更新出口监控配置
 */
export async function egressMonitorUpdateConfig(
  config: EgressMonitorConfig,
): Promise<void> {
  await invoke('egress_monitor_update_config', { config })
}

/**
 * 启动出口监控
 */
export async function egressMonitorStart(): Promise<void> {
  await invoke('egress_monitor_start')
}

/**
 * 停止出口监控
 */
export async function egressMonitorStop(): Promise<void> {
  await invoke('egress_monitor_stop')
}

/**
 * 获取出口监控统计
 */
export async function egressMonitorGetStats(): Promise<EgressMonitorStats> {
  return await invoke<EgressMonitorStats>('egress_monitor_get_stats')
}

/**
 * 重置出口监控统计
 */
export async function egressMonitorResetStats(): Promise<void> {
  await invoke('egress_monitor_reset_stats')
}

/**
 * 手动探测出口 IP
 */
export async function egressMonitorProbeNow(): Promise<EgressIpProbeResult> {
  return await invoke<EgressIpProbeResult>('egress_monitor_probe_now')
}

/**
 * 查询出口监控是否运行中
 */
export async function egressMonitorIsRunning(): Promise<boolean> {
  return await invoke<boolean>('egress_monitor_is_running')
}

File: src-tauri/src/cmd/egress_monitor.rs
use super::{CmdResult, StringifyErr};
use crate::core::egress_monitor::{EgressIpProbeResult, EgressMonitorConfig, EgressMonitorStats};

/// 获取出口监控配置
#[tauri::command]
pub async fn egress_monitor_get_config() -> CmdResult<EgressMonitorConfig> {
    Ok(crate::core::coordinator::get_coordinator()
        .get_advanced_config()
        .egress_monitor)
}

/// 更新出口监控配置
#[tauri::command]
pub async fn egress_monitor_update_config(config: EgressMonitorConfig) -> CmdResult<()> {
    crate::core::coordinator::update_advanced_config(move |advanced| {
        advanced.egress_monitor = config;
    })
    .await
    .stringify_err()
}

/// 启动出口监控
#[tauri::command]
pub async fn egress_monitor_start() -> CmdResult<()> {
    crate::core::coordinator::update_advanced_config(|advanced| {
        advanced.egress_monitor.enabled = true;
    })
        .await
        .stringify_err()
}

/// 停止出口监控
#[tauri::command]
pub async fn egress_monitor_stop() -> CmdResult<()> {
    crate::core::coordinator::update_advanced_config(|advanced| {
        advanced.egress_monitor.enabled = false;
    })
        .await
        .stringify_err()
}

/// 获取出口监控统计
#[tauri::command]
pub async fn egress_monitor_get_stats() -> CmdResult<EgressMonitorStats> {
    Ok(crate::core::egress_monitor::egress_monitor().get_stats())
}

/// 重置出口监控统计
#[tauri::command]
pub async fn egress_monitor_reset_stats() -> CmdResult<()> {
    crate::core::egress_monitor::egress_monitor().reset_stats();
    Ok(())
}

/// 手动探测出口 IP
#[tauri::command]
pub async fn egress_monitor_probe_now() -> CmdResult<EgressIpProbeResult> {
    crate::core::egress_monitor::egress_monitor()
        .probe_now()
        .await
        .stringify_err()
}

/// 查询出口监控是否运行中
#[tauri::command]
pub async fn egress_monitor_is_running() -> CmdResult<bool> {
    Ok(crate::core::egress_monitor::egress_monitor().is_running())
}
